### PR TITLE
Add GitHub Actions script

### PR DESCRIPTION
**Pull request by ghost**
**Workflow** 
It runs unit-tests for every commit on Ubuntu/macOS/Windows, this only take about 1 minute.

If the head commit message contains string `<release>`, it builds Python3.6~3.10 wheels, and upload to PyPI.
This will take more than 20 minutes, building `aarch64`/`ppc64le`/`s390x` wheels is slow, it uses `QEMU` to emulate these instruction sets on `x86_64` hosts.

If you don’t like this workflow, you can bring it up.

**x86/x86_64 Linux wheels**
For x86/x86_64 Linux wheels, it builds 3 types:
- manylinux1_i686
- manylinux1_x86_64
- manylinux2014_x86_64

This is because `manylinux2014_x86_64` build uses a new `memcpy()` function (`memcpy@GLIBC_2.14`). But in order to be compatible as much as possible, `manylinux1_x86_64` wheels are also built.

**Others**
I removed `PyPy` wheels, since there are 5~8 unit-tests failures.
We can add `PyPy` wheels later, and skip the failed unit-tests, this is easy.

Finally, please apply for a PyPI token, and create a `PYPI_TOKEN` variable in the GitHub repository. The value is the applied token.

I tested, it builds these wheels:
```
`build_wheels` job builds:
cp310-win32.whl
cp310-win_amd64.whl
cp36m-win32.whl
cp36m-win_amd64.whl
cp37m-win32.whl
cp37m-win_amd64.whl
cp38-win32.whl
cp38-win_amd64.whl
cp39-win32.whl
cp39-win_amd64.whl

cp310-macosx_10_9_x86_64.whl
cp310-macosx_11_0_arm64.whl
cp36m-macosx_10_9_x86_64.whl
cp37m-macosx_10_9_x86_64.whl
cp38-macosx_10_9_x86_64.whl
cp38-macosx_11_0_arm64.whl
cp39-macosx_10_9_x86_64.whl
cp39-macosx_11_0_arm64.whl

cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
cp310-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl
cp36m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
cp36m-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl
cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
cp37m-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl
cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
cp38-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl
cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
cp39-manylinux_2_5_i686.manylinux1_i686.manylinux_2_17_i686.manylinux2014_i686.whl

`build_in_manylinux2010` job builds:
cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl
cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl
cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl
cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl
cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl

`build_arch_wheels` job builds:
cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
cp36m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl

cp310-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl
cp36m-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl
cp37m-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl
cp38-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl
cp39-manylinux_2_17_ppc64le.manylinux2014_ppc64le.whl

cp310-manylinux_2_17_s390x.manylinux2014_s390x.whl
cp36m-manylinux_2_17_s390x.manylinux2014_s390x.whl
cp37m-manylinux_2_17_s390x.manylinux2014_s390x.whl
cp38-manylinux_2_17_s390x.manylinux2014_s390x.whl
cp39-manylinux_2_17_s390x.manylinux2014_s390x.whl
```

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# Automatically build binary wheels and source packages.
+name: cibuildwheel
+
+# Build on every branch push, tag push, and pull request change.
+on: [push, pull_request]
+
+env:
+  CIBW_BUILD: cp36-* cp37-* cp38-* cp39-* cp39-* cp310-*
+  CIBW_SKIP: "*musllinux*"
```
Comment (ghost) [inline .github/workflows/main.yml:9]: Without this, it also builds `musllinux` wheels. It seems unnecessary.

**Comment by ghost** [comment]
Another workflow is only build & upload when a certain branch is pushed, this way no string `<release>` in commit messages.

**Comment by mrabarnett** [comment]
How do I set the PYPI_TOKEN variable?

**Comment by ghost** [comment]
On `regex` repository, Setting->Secrets->New repository secret.

**Comment by ghost** [comment]
Seeing this modification, it's better to add a constraint, otherwise creating a PR may also trigger the heavily builds, don't know if it will also upload to PyPI.

For example, check the commit author, pushed branch, or else.
```diff
-    if: github.event_name == 'push' && contains(github.event.head_commit.message, '<release>')
+    if: github.event_name == 'push'
```

**Comment by mrabarnett** [comment]
Or should I have:
`on: [push]
`
so that it builds and uploads to PyPI when I push to it?
I only ever push when I want make a release.

**Comment by ghost** [comment]
> so that it builds and uploads to PyPI when I push to it?

Yes, but this workflow will try to release a new version for every commit, which may be a little inconvenient for development.

**Comment by mrabarnett** [comment]
Thanks. Your help has been invaluable.

**Comment by mrabarnett** [comment]
Somebody has just asked for muslllinux, so do I just need to remove:

    CIBW_SKIP: "*musllinux*"

?

**Comment by ghost** [comment]
https://www.python.org/dev/peps/pep-0656/

It seems that line can be removed.
I didn't know much about `musl` before this.

**Pull request merged**